src/Item.cpp

```diff
@@ -1,15 +1,50 @@
 #include "Item.hpp"
 
-Item::Item(const std::string &nm, const std::string &cat, const std::string &sub_cat, unsigned long qty,
-           unsigned long idd, double price, double cost, double tx)
-    : name(nm),
-      category(cat),
-      sub_category(sub_cat),
-      quantity(qty),
-      id(idd),
-      sale_price(price),
-      buy_cost(cost),
-      tax(tx) {
+Item::Item(const std::string &nm, const std::string &cat, const std::string &sub_cat, const std::string &qty,
+           const std::string &idd, const std::string &price, const std::string &cost, const std::string &tx) {
+    /* Validate input */
+    setValue("name", nm);
+    setValue("category", cat);
+    setValue("sub_category", sub_cat);
+    setValue("quantity", qty);
+    setValue("id", idd);
+    setValue("sale_price", price);
+    setValue("buy_cost", cost);
+    setValue("tax", tx);
+
```

Comment:
If an item is failed to be created as input do we want to create it with all the input that is valid or not add the item at all?
Unless I am mistaken, here we set everything and if it fails to set we throw an exception and then just move onto the next variable to set.

Reply:
The current behavior is:

- Item ctor is called.
- setValue() is called within the ctor.
- setValue() throws an exception if the value was invalid.
- This exception is caught wherever the Item was created / updated.

So what this results in is:
If I'm creating an Item with all bad input, it's going to fail on the ```setValue("quantity", qty);``` line.

This will throw an exception which will be caught where the item was created.

Since we're using smart pointers, we don't have to worry about deleting the Item object that failed to construct since it'll be destroyed when it goes out of scope.